### DISCUSSION TITLE
Stoch & KDJ

### DISCUSSION THREAD

**Discussion by muteddisk** [DaveSkender/Stock.Indicators#315] [Ideas]
Hi,
Good work on your code. Would you mind adding KDJ to your indicators? KDJ is Stoch with added J. The J line represents the divergence of the D value from the K value

thanks

**Comment by DaveSkender** [comment]
If %J really is just a weighted difference between [Stochastic Oscillator](https://daveskender.github.io/Stock.Indicators/indicators/Stoch/#content) %K and %D, I'll likely just add it to the output of Stoch and make reference to [KDJ](https://www.researchgate.net/publication/334896129_Mathematical_Analysis_Method_for_Stock_Market_Using_MA_and_KDJ_Indicator) in the documentation.  This is a similar technique used in MACD and PVO.  In the meantime, here's a suggestion on how to get %J in your wrapper code.

```csharp
// your custom derived class
public class KdjResult : StochResult
{
  public decimal PercentJ { get; set; }      // %J
}

[ .. ]

List<KdjResult> results = Indicator.GetStoch(history,9,3,3);
  .Select(x => new KdjResult
    {
      Date = x.Date,
      Oscillator = x.Oscillator,              // %K
      Signal = x.Signal,                      // %D
      PercentJ = 3*x.Oscillator - 2*Signal    // %J
    })
  .ToList();
```

**Comment by DaveSkender** [comment]
I'm looking for good references for the formula and I see most are using the standard Slow Stochastic values 9,3,3 as in my example above, then use `%J=3*%K-2*%D` or `%J=3*%D-2*%K` -- not sure which is correct yet.  I'm also wondering if those 3 and 2 weights should be derived from those smoothing factors or if they are truly static values?  I suspect they are static.  🤔 

If anyone has any legitimate documentation on this from the original author, George Lane, please share.

**Reply by DaveSkender** [reply]
I'm seeing both of these %J calculations used in various charting platforms on the web.  I've not found the original author's formula, because I can't find it (please share if you do); however, I'm going with `%J=3*%K-2*%D` because it's consistently depicted this way in a couple of academic research articles that I've found.  I also suspect this is correct since it depicts more of a leading indicator.

For comparison, here's `%J=3*%D-2*%K` (incorrect?):
![image](https://user-images.githubusercontent.com/8432125/104139646-fb2bcf80-537a-11eb-89eb-e6a14d05ccc0.png)

And here's `%J=3*%K-2*%D` (this is the one I'm using):
![image](https://user-images.githubusercontent.com/8432125/104139769-a63c8900-537b-11eb-959d-021a5d556f3e.png)

**Comment by DaveSkender** [comment]
Does anyone know how to trade this indicator?  Please explain if you're experienced with it -- do not just repost information and links from random blogs.  I'm curious to know how this is different from the simple %K / %D crossover signaling model.

**Comment by DaveSkender** [comment]
This is now available in version [1.6.3](https://github.com/DaveSkender/Stock.Indicators/releases/tag/1.6.3) for the `GetStoch` method.  See [Stochastic Oscillator](https://daveskender.github.io/Stock.Indicators/indicators/Stoch/#content) for more information.

**Comment by muteddisk** [comment]
Thanks bud your the best.

**Comment by wclark17** [comment]
Just curious if you could add an alias of K & D to Oscillator and Signal?  Totally not important, but helps my brain at times. 

BTW, You are saving me a ton of work elsewhere.  Happy to sponsor you and will increase it soon!

**Reply by DaveSkender** [reply]
I’ll take a look, track on #569.  I’ve had this same conversation in my head a few times, both ways, “what is D again?”  I settled on the more descriptive variants, but can see how others like the other naming.

let me see if I can add aliases in a way that doesn’t cause more potential confusion.

**Reply by DaveSkender** [reply]
and thank you for your patronage, much appreciated
